pkg/apis/core/validation/validation.go

```diff
@@ -3557,8 +3557,8 @@ func validatePodAffinityTerm(podAffinityTerm core.PodAffinityTerm, fldPath *fiel
 
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(podAffinityTerm.LabelSelector, fldPath.Child("matchExpressions"))...)
 	for _, name := range podAffinityTerm.Namespaces {
-		for _, msg := range ValidateNamespaceName(name, false) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), name, msg))
+		if _, msg := regexp.Compile(name); msg != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), name, msg.Error()))
```

Comment:
Affinity evaluation had significant performance problems when it was introduced. Any expansion of expressiveness to this feature needs to consider performance impact and describe how we will avoid regressions here.

Is there an accepted KEP for adding this functionality?

Reply:
We significantly improved affinity performance over the past four releases, we have benchmarks that test worst case scenarios (all pods in the same namespace, and all have affinity against/to each other) and show comparable scheduling throughput to vanilla workloads (100pod/sec).

I think adding a namespaceSelector is useful for cases like enforcing 1:1 pod per node across namespces (pods have required anti-affinity against each other on hostname topology), another is binpacking (pods across namespces have preferred affinity to each other).